src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -277,113 +277,57 @@ private void BoundsCheck(in SequencePosition start, in SequencePosition end)
-            Debug.Assert(End.GetObject() != null);
-            Debug.Assert(End.GetObject() is ReadOnlySequenceSegment<T>);
-
-            startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(Start.GetObject());
-            endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(End.GetObject());
+            startSegment = (ReadOnlySequenceSegment<T>)startObject;
+            startIndex = GetIndex(_sequenceStart);
+            endSegment = (ReadOnlySequenceSegment<T>)_sequenceEnd.GetObject();
+            endIndex = GetIndex(_sequenceEnd);
             return true;
         }
 
         internal bool TryGetArray(out ArraySegment<T> segment)
         {
-            GetTypeAndIndices(Start.GetInteger(), End.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
-            if (type != SequenceType.Array)
+            if (GetSequenceType() != SequenceType.Array)
             {
                 segment = default;
                 return false;
             }
 
-            Debug.Assert(Start.GetObject() != null);
-            Debug.Assert(Start.GetObject() is T[]);
-
-            segment = new ArraySegment<T>(Unsafe.As<T[]>(Start.GetObject()), startIndex, endIndex - startIndex);
-            return true;
-        }
-
-        internal bool TryGetReadOnlyMemory(out ReadOnlyMemory<T> memory)
-        {
-            GetTypeAndIndices(Start.GetInteger(), End.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
-
-            object startObject = Start.GetObject();
-            object endObject = End.GetObject();
-
-            Debug.Assert(startObject != null);
-            Debug.Assert(endObject != null);
-
-            int length = endIndex - startIndex;
-            if (startObject != endObject)
-            {
-                // Can't get ReadOnlyMemory from multi-block segments
-                memory = default;
-                return false;
-            }
-            else if (type == SequenceType.Array)
-            {
-                Debug.Assert(startObject is T[]);
-
-                memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
-            }
-            else if (type == SequenceType.MemoryManager)
-            {
-                Debug.Assert(startObject is MemoryManager<T>);
-
-                memory = Unsafe.As<MemoryManager<T>>(startObject).Memory;
-            }
-            else if (typeof(T) == typeof(char) && type == SequenceType.String)
-            {
-                Debug.Assert(startObject is string);
-
-                var text = Unsafe.As<string>(startObject);
-                memory = (ReadOnlyMemory<T>)(object)text.AsMemory();
-            }
-            else // ReadOnlySequenceSegment
-            {
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-
-                memory = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject).Memory;
-            }
+            Debug.Assert(_sequenceStart.GetObject() != null);
```

Comment:
Same here. I think we should handle this case rather than adding a debug.assert.

Reply:
For correct sequence `if (GetSequenceType() != SequenceType.Array)` above protect us from default sequence. So Assert only for sequences which were spoiled by threads race.

Reply:
> For correct sequence `if (GetSequenceType() != SequenceType.Array)` above protect us from default sequence.

Good point. Looks good to me. Thanks.